tests/integration/test_cogs_rates.py

```diff
@@ -0,0 +1,52 @@
+import os
+
+import pytest
+
+from poglink.cogs.rates import Rates
+from poglink.error import RatesFetchError
+from poglink.models.rates import RatesDiffItem
+
+
+@pytest.fixture()
+def rates_cog(sample_bot):
+    rates_cog = Rates(sample_bot)
+    yield rates_cog
+
+    # Clean up
+    for f in rates_cog.output_paths:
+        try:
+            os.remove(f)
+        except:
+            pass
+
+
+@pytest.mark.asyncio
+async def test_rates_compare_posted_rates(rates_cog, rates_url_1, rates_url_2):
+    output_path = rates_cog.output_paths[0]
+
+    # First request; no changes expected
+    rates_diff = await rates_cog.compare_posted_rates(rates_url_1, output_path)
+    assert len(rates_diff.items) == 0
+
+    # 2nd request; changes should be returned
+    rates_diff = await rates_cog.compare_posted_rates(rates_url_2, output_path)
+
+    assert len(rates_diff.items) == 3
+    assert (
+        RatesDiffItem(key="HarvestAmountMultiplier", old_val="3.0", new_val="2.0")
+        in rates_diff.items
+    )
+    assert (
+        RatesDiffItem(key="MatingIntervalMultiplier", old_val="0.6", new_val="0.7")
+        in rates_diff.items
+    )
+    assert (
+        RatesDiffItem(key="HexagonRewardMultiplier", old_val="1.5", new_val="2.5")
+        in rates_diff.items
+    )
+
+    # 3rd request; error
+    with pytest.raises(RatesFetchError):
+        rates_diff = await rates_cog.compare_posted_rates(
+            "http://localhost:5000/bogus-url.txt", output_path
+        )
```

Comment:
I think this is a good pattern for an integration test unless you want to also test the embed sending part (which might be fesible; I still want to look into [this](https://github.com/DXsmiley/dismock) (or hopefully something else if there's another variation of this library that someone has fleshed out more).

This at least tests the "check webpage, cache results, determine what to do" steps, which is more of an integration test than what we could do before. Having the discord embed part tested eventually would make this a full integration test that would be really helpful.

Also, maybe it's a bit lazy of me, but I didn't bother mocking up any webpage fixtures that would trigger the other custom exceptions, but those would be a welcome addition at some point to improve test coverage.

Reply:
Update: [this](https://pypi.org/project/distest/) project actually says it is an evolution of the previously linked pytest plugin.

Reply:
Update again: I decided to make more assertions because I realized it wasn't as hard as I first thought it would be

Reply:
#43